src/main/resources/db/migration/migrate_data_from_director_v1.sql

```diff
@@ -0,0 +1,155 @@
+on et.ecu_serial = e.ecu_serial AND et.namespace = e.namespace
+on duplicate key update
+  device_id = e.device,
+  public_key = e.public_key,
+  hardware_identifier = e.hardware_identifier,
+  current_target = et.id,
+  created_at = e.created_at, updated_at = e.updated_at
+;
+
+-- select count(*) from director.ecus;
+-- select count(*), ci.ecu_serial is null from director.ecus e left join director.current_images ci USING (namespace, ecu_serial) group by 2;
+-- select count(*), current_target is null from director.ecus e join director.current_images ci ON (namespace, ecu_serial) group by 2;
+
+-- devices
+insert into director_v2.devices (namespace, id, primary_ecu_id, created_at, updated_at)
+select namespace, device, ecu_serial, created_at, updated_at
+from director.ecus e
+where e.primary = 1
+on duplicate key update primary_ecu_id = e.ecu_serial, created_at = e.created_at, updated_at = e.updated_at;
+
+-- select count(*) from director.ecus where `primary` = 1 ;
+
+INSERT into director_v2.hardware_updates (namespace, id, hardware_identifier, to_target_id, from_target_id, target_format, created_at, updated_at)
+select to_mtu.namespace, to_mtu.id, to_mtu.hardware_identifier, to_mtu.target_id, from_mtu.target_id, to_mtu.target_format, to_mtu.created_at, to_mtu.updated_at
+FROM
+(select mtu.namespace, mtu.id, mtu.hardware_identifier, et.id target_id, target_format, mtu.created_at, mtu.updated_at
+  from director.multi_target_updates mtu JOIN director_v2.ecu_targets_v1 et ON mtu.namespace = et.namespace AND mtu.target = et.filename AND mtu.target_hash = et.sha256 AND mtu.target_size = et.length) to_mtu
+LEFT JOIN
+(select mtu.namespace, mtu.id, mtu.hardware_identifier, et.id target_id from director.multi_target_updates mtu JOIN director_v2.ecu_targets_v1 et ON mtu.namespace = et.namespace AND mtu.from_target = et.filename AND mtu.from_target_hash = et.sha256 AND mtu.from_target_size = et.length) from_mtu
+ON to_mtu.namespace = from_mtu.namespace AND to_mtu.id = from_mtu.id AND to_mtu.hardware_identifier = from_mtu.hardware_identifier
+ON duplicate key update target_format = VALUES(target_format), updated_at = VALUES(updated_at)
+;
+
+-- select count(*), from_target is null  from director.multi_target_updates group by 2;
+-- select count(*), from_target_id is null from hardware_updates group by 2;
+
+
+DROP TEMPORARY TABLE IF EXISTS assignments_v1;
+create temporary table assignments_v1 AS
+select
+  eua.namespace, eua.device_id, eua.ecu_id ecu_serial, et.id ecu_target_id, dua.correlation_id, dua.served in_flight, eua.created_at, eua.updated_at,
+   (ranked_eua.rank = 1 AND dct.device_current_target < eua.version) running,
+   eua.version ecu_version, ranked_eua.rank version_rank, dct.device_current_target
+FROM director.ecu_update_assignments eua
+JOIN
+  (select namespace, ecu_id, version, device_id, ROW_NUMBER() OVER (PARTITION BY device_id ORDER BY version DESC) rank from director.ecu_update_assignments) ranked_eua
```

Comment:
So this will produce tuples like `(ns, ecu, version, did1, 1), (ns, ecu, version, did1, 2), (ns, ecu, version, did2, 1), (ns, ecu, version, did2, 2), ...`, correct?
And what's the meaning of `(ranked_eua.rank = 1 AND dct.device_current_target < eua.version) running` then? Is it to chose the latest target assigned to that ECU? Or the one that it's installed? I'm not sure I follow here...

Reply:
Yes produces tuples like that, wher version is decreasing, so when rank = 1, and `t.device_current_target < eua.version)` it means the device has reported a targets.json version that is lower than what is assigned to it, meaning the assignment is still in process and not yet installed.